processing/src/main/java/io/druid/query/groupby/epinephelinae/SpillingGrouper.java

```diff
@@ -0,0 +1,198 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Function;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import com.metamx.common.guava.CloseQuietly;
+import com.metamx.common.logger.Logger;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.segment.ColumnSelectorFactory;
+import net.jpountz.lz4.LZ4BlockInputStream;
+import net.jpountz.lz4.LZ4BlockOutputStream;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Grouper based around a single underlying {@link BufferGrouper}. Not thread-safe.
+ *
+ * When the underlying grouper is full, its contents are sorted and written to temporary files using "spillMapper".
+ */
+public class SpillingGrouper<KeyType extends Comparable<KeyType>> implements Grouper<KeyType>
+{
+  private static final Logger log = new Logger(SpillingGrouper.class);
+
+  private final BufferGrouper<KeyType> grouper;
+  private final KeySerde<KeyType> keySerde;
+  private final LimitedTemporaryStorage temporaryStorage;
+  private final ObjectMapper spillMapper;
+  private final AggregatorFactory[] aggregatorFactories;
+
+  private final List<File> files = Lists.newArrayList();
+  private final List<Closeable> closeables = Lists.newArrayList();
+
+  public SpillingGrouper(
+      final ByteBuffer buffer,
+      final KeySerdeFactory<KeyType> keySerdeFactory,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final LimitedTemporaryStorage temporaryStorage,
+      final ObjectMapper spillMapper,
+      final int maxBufferGrouperSize
+  )
+  {
+    this.keySerde = keySerdeFactory.factorize();
+    this.grouper = new BufferGrouper<>(
+        buffer,
+        keySerde,
+        columnSelectorFactory,
+        aggregatorFactories,
+        maxBufferGrouperSize
+    );
+    this.aggregatorFactories = aggregatorFactories;
+    this.temporaryStorage = temporaryStorage;
+    this.spillMapper = spillMapper;
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    if (grouper.aggregate(key, keyHash)) {
+      return true;
+    } else {
+      // TODO(gianm): This blocks up all other processing threads, maybe split the buffer so we can spill in background
```

Comment:
`spill()` occurs in a processing thread, and can take a while (tens of seconds). While this is happening, no other processing thread for the same query can make progress. If all processing threads on the node are currently working on that query, then the entire pool can be jammed up until the spill is done.

I think this could be handled better in the future but wasn't planning to do better for this particular PR. People that really hate this behavior could disable disk spilling by setting `druid.query.groupBy.maxOnDiskStorage = 0`.

The comment about "split the buffer" refers to potentially handling this better by having two buffers, so we can spill one in the background while accumulating into a new one (like RT indexing does). That would help a bit but would still potentially be problematic if we can't spill as fast as we can accumulate. More creative solutions might be needed too.